datafusion/functions/src/math/random.rs

```diff
@@ -64,45 +61,9 @@ impl ScalarUDFImpl for RandomFunc {
         Ok(Float64)
     }
 
-    fn invoke(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
-        random(args)
-    }
-}
-
-/// Random SQL function
-fn random(args: &[ColumnarValue]) -> Result<ColumnarValue> {
-    let len: usize = match &args[0] {
-        ColumnarValue::Array(array) => array.len(),
-        _ => return exec_err!("Expect random function to take no param"),
-    };
-    let mut rng = thread_rng();
-    let values = iter::repeat_with(|| rng.gen_range(0.0..1.0)).take(len);
-    let array = Float64Array::from_iter_values(values);
-    Ok(ColumnarValue::Array(Arc::new(array)))
-}
-
-#[cfg(test)]
-mod test {
-    use std::sync::Arc;
-
-    use arrow::array::NullArray;
-
-    use datafusion_common::cast::as_float64_array;
-    use datafusion_expr::ColumnarValue;
-
-    use crate::math::random::random;
-
-    #[test]
-    fn test_random_expression() {
-        let args = vec![ColumnarValue::Array(Arc::new(NullArray::new(1)))];
-        let array = random(&args)
-            .expect("failed to initialize function random")
-            .into_array(1)
-            .expect("Failed to convert to array");
-        let floats =
-            as_float64_array(&array).expect("failed to initialize function random");
-
-        assert_eq!(floats.len(), 1);
-        assert!(0.0 <= floats.value(0) && floats.value(0) < 1.0);
+    fn invoke(&self, _args: &[ColumnarValue]) -> Result<ColumnarValue> {
```

Comment:
I think the idea here is that expectation is that `rand` is invoked *once per row* rather than *once per batch*. And the only way it knew how many rows to make is to get a null array in 🤔 

For example, when I run `datafusion-cli` from this PR to call random() the same value is returned for each row:

```sql
> create table foo as values (1), (2), (3), (4), (5);
0 row(s) fetched.
Elapsed 0.018 seconds.

> select column1, random() from foo;
+---------+--------------------+
| column1 | random()           |
+---------+--------------------+
| 1       | 0.9594375709000513 |
| 2       | 0.9594375709000513 |
| 3       | 0.9594375709000513 |
| 4       | 0.9594375709000513 |
| 5       | 0.9594375709000513 |
+---------+--------------------+
5 row(s) fetched.
Elapsed 0.012 seconds.
```

But I expect that each row has a different value for `random()`

However, since none of the tests failed, clearly we have a gap in test coverage 🤔

Reply:
Nice catch! Let me think about how to design it, I would prefer something like `support_random` to specialize `random()` case.

Reply:
What about adding a `invoke_no_args(num_rows: usize)` method to the `ScalarUDFImpl` -- with a default implementation that returns "not implemented" error 

That might make it clear what was happening and would provide clear semantics about what to do in this case 🤔

Reply:
I didn't see the message, it is also a good idea.

Reply:
Do we test the `uuid` function? i think the `uuid` has the same attribute like `random`
cc @jayzhan211

Reply:
We have test like 
```
query II
SELECT octet_length(uuid()), length(uuid())
----
36 36
```

Reply:
> We have test like
> 
> ```
> query II
> SELECT octet_length(uuid()), length(uuid())
> ----
> 36 36
> ```

I means the `uuid` function is also `invoked once per row rather than once per batch` like `random()` function mentioned by alamb.

I test the function in the spark

```
spark-sql> desc test;
col1    int     NULL
Time taken: 0.065 seconds, Fetched 1 row(s)


spark-sql> select * from test;
1
2



spark-sql> select *,uuid() from test;
1       6b04b66c-2e6c-4925-8b18-a9d51d5ed80a
2       3e0be0c2-9ff2-422a-8f3f-5cdb6551264b
```

Reply:
file a issue for this: https://github.com/apache/datafusion/issues/10247